libcontainer/specconv/spec_linux.go

```diff
@@ -650,13 +654,29 @@ func stringToDeviceRune(s string) (configs.DeviceType, error) {
 	}
 }
 
-func createDevices(spec *specs.Spec, config *configs.Config) error {
-	// Add default set of devices.
-	for _, device := range AllowedDevices {
-		if device.Path != "" {
-			config.Devices = append(config.Devices, device)
+func createDevices(spec *specs.Spec, config *configs.Config) ([]*configs.Device, error) {
+
+	// If a spec device is redundant with a default device, remove that default
+	// device (the spec one takes priority).
+	dedupedAllowDevs := []*configs.Device{}
+
+	if spec.Linux != nil {
+		for _, ad := range AllowedDevices {
+			conflict := false
+			for _, sd := range spec.Linux.Devices {
+				if sd.Path == ad.Path {
+					conflict = true
+				}
+			}
+			if !conflict {
```

Comment:
I would use
```go
if conflict {
  break
}
dedupedAllowDevs = append(dedupedAllowDevs, ad)
....
```

here since it's more readable

Reply:
Of even
```go
next:
for _, ad := range AllowedDevices {
  for _, sd := range spec.Linux.Devices {
    if sd.Path == ad.Path {
      continue next
    }
  }
  dedupedAllowDevs = append(dedupedAllowDevs, ad)
  if ad.Path != "" {
    config.Devices = append(config.Devices, ad)
  }
}
```

Reply:
Maybe
```go
next:
for _, ad := range AllowedDevices {
  if ad.Path != "" {
    for _, sd := range spec.Linux.Devices {
      if sd.Path == ad.Path {
        continue next
      }
    }
  }
  dedupedAllowDevs = append(dedupedAllowDevs, ad)
  if ad.Path != "" {
    config.Devices = append(config.Devices, ad)
  }
}
```

as otherwise, if there's an entry in spec.Linux.Devices with Path = "", we'll skip all the AllowDevices with empty path as well, and this seems wrong. It seems your code has the same problem.

Reply:
That looks nice ... thanks. Let me re-work this section of the code once more per your suggestion.

Reply:
Hi @kolyshkin, I followed your suggestion and re-worked the loop. Thanks!